### DISCUSSION TITLE
Rule with optional suffix, suffix must end in a specific character

### DISCUSSION THREAD

**Discussion by filippodebortoli** [pest-parser/pest#742] [Q&A] [answered]
I am currently trying to encode the following rule using pest (an extremely simplified version of [this](https://www.w3.org/TR/2008/REC-rdf-sparql-query-20080115/#rPN_PREFIX)):

````
ASCII_ALPHA ((ASCII_ALPHANUMERIC|'.')* ASCII_ALPHANUMERIC)?
````

My pest rule looks as follows:

````
test_rule = @{ASCII_ALPHA ~ ((ASCII_ALPHANUMERIC|".")* ~ ASCII_ALPHANUMERIC)? }
````
My expectation is that `MyParser::parse(Rule::test_rule, "test")` should execute successfully and consume the entire string.
What is actually happening is that the parser only matches the first character `"t"`.
By poking around the rule, I found out that this is the one part that prevents this parser from matching the whole string:
````
test_rule = @{ASCII_ALPHA ~ ((ASCII_ALPHANUMERIC|".")* ~ ASCII_ALPHANUMERIC)? }
                                                       ^------------------^
````
Have I encountered a bug, or is there a specific reason this rule is not working as I am expecting?

I thank you very much for this fantastic crate and the hard work you put into it! :-)

**Comment by birkenfeld** [comment, marked as answer]
I think the problem here is that matching is greedy and does not backtrack. I.e. all the `est` are consumed in the `(ASCII_ALPHANUMERIC|".")*` part and there is nothing left for the final `ASCII_ALPHANUMERIC`. Therefore, the whole `(...)?` part fails.

I think this rule should work instead:
```
test_rule = @{ASCII_ALPHA ~ ("."* ~ ASCII_ALPHANUMERIC)* }
```

**Reply by filippodebortoli** [reply]
The rule you are suggesting works and seems to be equivalent to the one I need (after a quick check).
Thank you very much for this nice reply and for explaining why the original rule does not work as intended!

**Reply by birkenfeld** [reply]
You're welcome! I opened #744 to make this clearer in the docs.
